### PR TITLE
Image and shape clip paths

### PR DESCRIPTION
**Pull request by alexcjohnson**
In testing axis reference changes for integrating v1.24 into the workspace, I noticed some strange situations with clip paths for shapes and images. Turns out they had different issues but related symptoms. There's one particular edge case that I didn't sort out here because it's a much bigger structural question, so I described it in #1452 instead.

@etpinard please review.

### CODE REVIEW

**Comment by alexcjohnson** [comment]
Note: annotations do not have these issues, as they don't use clip paths. We will happily allow the arrow and text box of an annotation to go outside the plot area as long as its reference point is visible.

File: src/components/images/draw.js [hunk 1]
```diff
@@ -143,9 +143,10 @@ module.exports = function draw(gd) {
             yId = ya ? ya._id : '',
             clipAxes = xId + yId;
 
-        if(clipAxes) {
-            thisImage.call(Drawing.setClipUrl, 'clip' + fullLayout._uid + clipAxes);
-        }
+        thisImage.call(Drawing.setClipUrl, clipAxes ?
+            ('clip' + fullLayout._uid + clipAxes) :
+            null
+        );
```
Comment (alexcjohnson) [inline src/components/images/draw.js]: Needed to *unset* the clip path if you change from data-referenced to fully paper-referenced.

File: test/jasmine/tests/shapes_test.js [hunk 2]
```diff
@@ -382,6 +382,71 @@ describe('Test shapes:', function() {
...
+                fillcolor: '#eee', layer: 'below'
+            }]
+        }).then(done);
+    });
+
+    afterEach(destroyGraphDiv);
+
+    function getShape(index) {
+        var s = d3.selectAll('path[data-index="' + index + '"]');
+        expect(s.size()).toBe(1);
```
Comment (alexcjohnson) [inline test/jasmine/tests/shapes_test.js]: before the fix above, the shape was drawn twice in its initial configuration `xref: 'x', yref: 'paper'`: once on subplot `xy` and again on `xy2`

File: src/components/shapes/draw.js [hunk 3]
```diff
@@ -110,10 +110,15 @@ function drawOne(gd, index) {
...
-                'clip' + gd._fullLayout._uid + clipAxes);
-        }
+        // note that for layer="below" the clipAxes can be different from the
+        // subplot we're drawing this in. This could cause problems if the shape
+        // spans two subplots. See https://github.com/plotly/plotly.js/issues/1452
+        var clipAxes = (options.xref + options.yref).replace(/paper/g, '');
+
+        path.call(Drawing.setClipUrl, clipAxes ?
+            ('clip' + gd._fullLayout._uid + clipAxes) :
+            null
```
Comment (etpinard) [inline src/components/shapes/draw.js]: Nice way to unset a clip path. :+1:

File: test/jasmine/tests/layout_images_test.js [hunk 4]
```diff
@@ -425,23 +425,38 @@ describe('images log/linear axis changes', function() {
         // we don't try to figure out the position on a new axis / canvas
         // automatically when you change xref / yref, we leave it to the caller.
 
+        // initial clip path should end in 'xy' to match xref/yref
+        expect(d3.select('image').attr('clip-path') || '').toMatch(/xy\)$/);
```
Comment (etpinard) [inline test/jasmine/tests/layout_images_test.js]: Ha. I didn't know about `.toMatch`. Pretty handy :hammer:

File: test/jasmine/tests/layout_images_test.js [hunk 5]
```diff
@@ -425,23 +425,38 @@ describe('images log/linear axis changes', function() {
...
             // log to paper
             return Plotly.relayout(gd, {'images[0].yref': 'paper'});
         })
         .then(function() {
             expect(gd.layout.images[0].y).toBe(1);
 
+            expect(d3.select('image').attr('clip-path') || '').toMatch(/x\)$/);
+
+            // change to full paper-referenced, to make sure the clip path disappears
+            return Plotly.relayout(gd, {'images[0].xref': 'paper'});
```
Comment (etpinard) [inline test/jasmine/tests/layout_images_test.js]: I'm surprised we weren't testing this before. Oh well, now it's :lock: ed

File: test/jasmine/tests/shapes_test.js (hunk 2 above)
Comment (etpinard) [inline test/jasmine/tests/shapes_test.js]: Good eyes. Thanks very much!

**Comment by etpinard** [comment]
Looks like `shapes_below_traces` produced a diff

image

is that on purpose?

**Comment by alexcjohnson** [comment]
> Looks like shapes_below_traces produced a diff, is that on purpose?

Hmm, maybe #1452 needs to be addressed together with this PR after all, that's what's going on here. You'll see that the data in the lower left subplot gets covered up in the new result for that image. But the original is clearly wrong in how it breaks the two shapes that span that subplot and either the one above or to the right into two disjoint pieces.

From a slack chat:

> the reason I said (in #1452 ) "I'm not sure this really has a solution even theoretically” is that there really isn’t a layering that puts the shape in front of the plot area backgrounds but behind the traces, at least not when you consider inset subplots that want to have their own backgrounds in front of the main plot traces.
>
> One easy thing we could do is put the below shapes completely below the plot layers… then you have to make `plot_bgcolor` transparent to see the shapes.
>
> A trickier thing we could do is pull the background rects out of the subplot layers, putting them behind everything else, UNLESS we detect overlapping (but not overlaying) subplots.
>
> That would cover all the use cases I can think of except below shapes that are half referenced to an axis of an inset subplot… which would be an exceedingly weird situation.

@etpinard suggested:
> Maybe we should disallow shapes that span multiple subplots

That would certainly make this easier (the restriction would only apply to `layer: 'below'` shapes and images) - then we'd just have to figure out which subplot contains the shape. But there are some cool uses of subplot-spanning shapes - for example on the standard finance chart that has candlesticks/OHLC in the main plot and volume in a little lower subplot, if you want to highlight a time range. It would look weird to force these to be two separate shapes. Would also be annoying to try and enforce this restriction during editing (particularly via drag)

So I'm leaning toward the "trickier thing": pulling the plot backgrounds back to their own layer except for inset plots.

Thoughts?

**Comment by alexcjohnson** [comment]
#1390 is wrapped up in this too... better fix that at the same time.

**Comment by etpinard** [comment]
> So I'm leaning toward the "trickier thing": pulling the plot backgrounds back to their own layer except for inset plots.

This is definitely the _best_ solution from a user's perspective. If you feel like taking a shot at this, go for it :+1:

**Comment by alexcjohnson** [comment]
https://github.com/plotly/plotly.js/pull/1453/commits/2102b0f39c9fa09a824417699598baeb9e045a83 fixes #1390 and #1452 - I still want to add a couple of tests of the inset vs flat background behavior though

File: src/components/images/draw.js [hunk 6]
```diff
@@ -151,29 +163,44 @@ module.exports = function draw(gd) {
...
     });
+
+    var allSubplots = Object.keys(fullLayout._plots);
+    for(i = 0; i < allSubplots.length; i++) {
+        subplot = allSubplots[i];
+        var subplotObj = fullLayout._plots[subplot];
+
+        // filter out overlaid plots (which havd their images on the main plot)
+        // and gl2d plots (which don't support below images, at least not yet)
+        if(!subplotObj.imagelayer) continue;
```
Comment (etpinard) [inline src/components/images/draw.js]: Looks good :+1:

File: src/plot_api/plot_api.js [hunk 7]
```diff
@@ -308,8 +308,7 @@ Plotly.plot = function(gd, data, layout, config) {
 
         // keep reference to shape layers in subplots
         var layerSubplot = fullLayout._paper.selectAll('.layer-subplot');
-        fullLayout._imageSubplotLayer = layerSubplot.selectAll('.imagelayer');
-        fullLayout._shapeSubplotLayer = layerSubplot.selectAll('.shapelayer');
+        fullLayout._shapeSubplotLayers = layerSubplot.selectAll('.shapelayer');
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: :pencil2: nicely done

File: test/jasmine/tests/cartesian_test.js [hunk 8]
```diff
@@ -340,6 +352,70 @@ describe('subplot creation / deletion:', function() {
...
+
+            // verify that these are *all* the subplots and backgrounds we have
+            expect(gd.querySelectorAll('.subplot').length).toBe(3);
+            ['xy', 'x2y2', 'xy3'].forEach(function(subplot) {
+                expect(gd.querySelectorAll('.subplot.' + subplot).length).toBe(1);
+            });
+            expect(gd.querySelectorAll('.bg').length).toBe(behindCount + x2y2Count);
+        }
+
+        Plotly.plot(gd, [
```
Comment (etpinard) [inline test/jasmine/tests/cartesian_test.js]: Fantastic test :tada:

**Comment by etpinard** [comment]
:dancer: nicely done!

Don't forget to close off  #1390 and #1452 after merging :beers:

**Pull request merged**